### PR TITLE
chore: Patch `revm` to current head commit

### PR DESCRIPTION
**Pull request by clabby**
## Overview

> **Note**
> WIP - waiting on a few changes to land in `revm` to pull out of draft and finalize.

Upgrades revm to a commit that includes https://github.com/bluealloy/revm/pull/789. Necessary for moving forwards with the `op-reth` upgrades in https://github.com/paradigmxyz/reth/pull/4377.

### TODO
- [x] Fix JS tracer fault. New `revm` API no longer passes the `eval`, how do we check if the result was a failure?
- [x] Remove `contract.rs` - the file is no longer needed with `create2_from_code` being implemented on `Address`.
- [ ] Patch crate rather than override crates.io dep.

### CODE REVIEW

File: crates/revm/revm-inspectors/src/tracing/js/mod.rs [hunk 1]
```diff
@@ -324,32 +327,31 @@ where
...
-                cost: interp.gas.spend(),
-                depth: data.journaled_state.depth(),
-                refund: interp.gas.refunded() as u64,
-                error: Some(format!("{:?}", eval)),
-                contract: self.active_call().contract.clone(),
-            };
-
-            let _ = self.try_fault(step, db);
-        }
+        // if matches!(eval, return_revert!()) {
```
Comment (clabby) [inline crates/revm/revm-inspectors/src/tracing/js/mod.rs:337]: Not immediately obvious how we're supposed to do this now that the result isn't passed. Any ideas @rkrasiuk? Might have missed something in the new API.

File: crates/revm/revm-inspectors/src/tracing/mod.rs [hunk 2]
```diff
@@ -371,7 +373,7 @@ impl TracingInspector {
         step.gas_cost = step.gas_remaining - self.gas_inspector.gas_remaining();
 
         // set the status
-        step.status = status;
+        step.status = InstructionResult::Continue;
```
Comment (clabby) [inline crates/revm/revm-inspectors/src/tracing/mod.rs:376]: Also not immediately obvious that this is correct. Previously we'd passed the state along here.

File: crates/revm/revm-inspectors/src/tracing/js/mod.rs [hunk 3]
```diff
@@ -206,6 +204,7 @@ impl JsInspector {
         )?)
     }
 
+    #[allow(unused)]
```
Comment (clabby) [inline crates/revm/revm-inspectors/src/tracing/js/mod.rs:207]: Will be removed when the JS tracer is fixed.

**Comment by clabby** [comment]
There is also a disallowed license within the `aurora-engine-modexp` dependency of this version of `revm` that was introduced in https://github.com/bluealloy/revm/pull/769. cc @rakita

**Comment by rakita** [comment]
> There is also a disallowed license within the `aurora-engine-modexp` dependency of this version of `revm` that was introduced in

Thanks for flagging this, there are a few CC0 licences that we use in reth, you can add it to the list here: https://github.com/paradigmxyz/reth/blob/a60dbfdd56e5fc2d5df8d87089952347008c1d39/deny.toml#L60-L70

File: crates/revm/revm-inspectors/src/tracing/js/mod.rs [hunk 4]
```diff
@@ -324,32 +327,31 @@ where
 
     fn step_end(
         &mut self,
-        interp: &mut Interpreter,
-        data: &mut EVMData<'_, DB>,
-        eval: InstructionResult,
+        _interp: &mut Interpreter<'_>,
+        _data: &mut EVMData<'_, DB>,
     ) -> InstructionResult {
```
Comment (mattsse) [inline crates/revm/revm-inspectors/src/tracing/js/mod.rs:332]: @rakita why was this removed, or has this been moved to the `Interpreter.instruction_result`?

Comment (rakita) [reply crates/revm/revm-inspectors/src/tracing/js/mod.rs:332]: It was the same thing, `interpreter.instruction_result` should be used

Comment (clabby) [reply crates/revm/revm-inspectors/src/tracing/js/mod.rs:332]: Nice, that answers my above q - thanks!

File: crates/revm/revm-inspectors/src/tracing/js/mod.rs (hunk 1 above)
Comment (clabby) [reply crates/revm/revm-inspectors/src/tracing/js/mod.rs:337]: Answered in https://github.com/paradigmxyz/reth/pull/5032#discussion_r1360770302; pending fix.

**Comment by mattsse** [comment]
closing this because I need to get this over the line and I'm unable to push to this branch directly

https://github.com/paradigmxyz/reth/pull/5109
